File: packages/protocol/src/events.ts
/**
 * Event types for the pub/sub event bus.
 */

import type { ExtensionStatus } from "./extensions";

/** Extension was loaded event. */
export interface ExtensionLoadedEvent {
  type: "extension_loaded";
  extensionId: string;
}

/** Extension was unloaded event. */
export interface ExtensionUnloadedEvent {
  type: "extension_unloaded";
  extensionId: string;
}

/** Extension status changed event. */
export interface ExtensionStatusChangedEvent {
  type: "extension_status_changed";
  extensionId: string;
  status: ExtensionStatus;
}

/** Process output from a runner event. */
export interface ProcessOutputEvent {
  type: "process_output";
  sessionId: string;
  data: string;
}

/** Process exited event. */
export interface ProcessExitedEvent {
  type: "process_exited";
  sessionId: string;
  exitCode?: number;
}

/** Configuration changed event. */
export interface ConfigChangedEvent {
  type: "config_changed";
  key: string;
}

/** Window visibility changed event. */
export interface WindowVisibilityChangedEvent {
  type: "window_visibility_changed";
  visible: boolean;
}

/** Events emitted by the application. */
export type AppEvent =
  | ExtensionLoadedEvent
  | ExtensionUnloadedEvent
  | ExtensionStatusChangedEvent
  | ProcessOutputEvent
  | ProcessExitedEvent
  | ConfigChangedEvent
  | WindowVisibilityChangedEvent;

/** Event subscription request. */
export interface EventSubscription {
  /** Event types to subscribe to (empty = all) */
  eventTypes: string[];
}

File: crates/protocol/src/events.rs
//! Event types for the pub/sub event bus.

use serde::{Deserialize, Serialize};

/// Events emitted by the application.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum AppEvent {
    /// Extension was loaded
    ExtensionLoaded { extension_id: String },
    /// Extension was unloaded
    ExtensionUnloaded { extension_id: String },
    /// Extension status changed
    ExtensionStatusChanged {
        extension_id: String,
        status: super::ExtensionStatus,
    },
    /// Process output from a runner
    ProcessOutput {
        session_id: String,
        data: String,
    },
    /// Process exited
    ProcessExited {
        session_id: String,
        exit_code: Option<i32>,
    },
    /// Configuration changed
    ConfigChanged { key: String },
    /// Window visibility changed
    WindowVisibilityChanged { visible: bool },
}

/// Event subscription request.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EventSubscription {
    /// Event types to subscribe to (empty = all)
    pub event_types: Vec<String>,
}
